### ISSUE TITLE
Wrong error message with closures when variable moved

### ISSUE THREAD

**Issue by shengsheng** [rust-lang/rust#63405]
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c4b92f189e40cc2a6e8a6d13a24e4ee8
```rust
struct T(u32);

fn by_value(_: T) {}
fn by_ref(_: &T) {}
    
fn main() {
    let z = T(123);
    
    let closure = || {
        by_ref(&z);
        by_value(z);
    };
    
    closure();
    
    by_value(z);
}
```

Output:
```
   Compiling playground v0.0.1 (/playground)
error[E0382]: use of moved value: `z`
  --> src/main.rs:16:14
   |
7  |     let z = T(123);
   |         - move occurs because `z` has type `T`, which does not implement the `Copy` trait
8  |     
9  |     let closure = || {
   |                   -- value moved into closure here
10 |         by_ref(&z);
   |                 - variable moved due to use in closure
...
16 |     by_value(z);
   |              ^ value used here after move

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

The `move` actually happen in line 11: `by_value(z);` not line 10.

Expect output:

```
11 |         by_value(z);
   |                  - variable moved due to use in closure
```

**Comment by shengsheng** [comment]
Another example:

```rust
#[derive(Debug)]
struct V(u32);

fn main() {
    let a = Some(456);
    let v = V(123);
    
    a.map(|x| {
        let _ = x + v.0;
        drop(v);
    });
    
    println!("v: {:?}", v);
}
```

Output:
```
   Compiling playground v0.0.1 (/playground)
error[E0382]: borrow of moved value: `v`
  --> src/main.rs:13:25
   |
6  |     let v = V(123);
   |         - move occurs because `v` has type `V`, which does not implement the `Copy` trait
7  |     
8  |     a.map(|x| {
   |           --- value moved into closure here
9  |         let _ = x + v.0;
   |                     - variable moved due to use in closure
...
13 |     println!("v: {:?}", v);
   |                         ^ value borrowed here after move

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by shengsheng** [comment]
Also confusing with `move` keyword:
```rust
#[derive(Debug)]
struct V(u32);

fn main() {
    let a = Some(456);
    let v = V(123);
    
    a.map(move |x| {
        let _ = x + v.0;
    });
    
    println!("v: {:?}", v);
}
```

Output:
```
   Compiling playground v0.0.1 (/playground)
error[E0382]: borrow of moved value: `v`
  --> src/main.rs:12:25
   |
6  |     let v = V(123);
   |         - move occurs because `v` has type `V`, which does not implement the `Copy` trait
7  |     
8  |     a.map(move |x| {
   |           -------- value moved into closure here
9  |         let _ = x + v.0;
   |                     - variable moved due to use in closure
...
12 |     println!("v: {:?}", v);
   |                         ^ value borrowed here after move

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by arora-aman** [comment]
This has been fixed on the nightly https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a0564c5e235b20202f28fea65af76e19
